returnn/datasets/multi_proc.py

```diff
@@ -87,61 +95,74 @@ def _meta_info_cache(self):
-            for i in range(self.num_workers - 1):
-                reader, writer = _mp.Pipe(duplex=False)
-                seq_order_to_worker.append(writer)
-                worker_from_seq_order.append(reader)
-
-            worker_parent_conns = []  # type: List[mpConnection]
-            worker_child_conns = []  # type: List[mpConnection]
-            for i in range(self.num_workers):
-                parent_conn, child_conn = _mp.Pipe()
-                worker_parent_conns.append(parent_conn)
-                worker_child_conns.append(child_conn)
-
-            worker_procs = []
-            for i in range(self.num_workers):
-                worker_proc = _mp.Process(
-                    name=f"{self.name} worker proc {i + 1}/{self.num_workers}",
-                    target=self._worker_proc_loop,
-                    args=(
-                        i,
-                        self.dataset,
-                        self.buffer_size,
-                        worker_child_conns[i],
-                        worker_from_seq_order[i - 1] if i > 0 else None,
-                        seq_order_to_worker if i == 0 else None,
-                    ),
-                    daemon=True,
-                )
-                worker_proc.start()
-                worker_procs.append(worker_proc)
-                # Make sure the child connection is closed here.
-                # It stays open in the child, until the child dies.
-                # When that happens, now any consecutive read on the pipe
-                # should yield an exception -- which is what we want,
-                # otherwise it would just hang.
-                worker_child_conns[i].close()
-
-            self._seq_order_proc_parent_conn = worker_parent_conns[0]  # type: mpConnection
-            self._worker_parent_conns = worker_parent_conns
-            self._worker_procs = worker_procs
-
-            self._seq_order_proc_parent_conn.send(("init", {}))
-            msg, self.num_inputs = self._seq_order_proc_parent_conn.recv()
-            assert msg == "num_inputs"
-            msg, self.num_outputs = self._seq_order_proc_parent_conn.recv()
-            assert msg == "num_outputs"
-            msg, self._total_num_seqs = self._seq_order_proc_parent_conn.recv()
-            assert msg == "total_num_seqs"
-            msg, self.labels = self._seq_order_proc_parent_conn.recv()
-            assert msg == "labels"
+        if self._worker_procs:
+            return
+
+        _mp = NonDaemonicSpawnContext(process_pre_init_func=SubProcCopyGlobalConfigPreInitFunc())
+
+        # Seq order proc (first worker) directly sends the seq order to each (other) worker.
+        seq_order_to_worker = []  # type: List[mpConnection]
+        worker_from_seq_order = []  # type: List[mpConnection]
+        for i in range(self.num_workers - 1):
+            reader, writer = _mp.Pipe(duplex=False)
+            seq_order_to_worker.append(writer)
+            worker_from_seq_order.append(reader)
```

Comment:
These are only needed for the `seq_order` sharding method, right?

Reply:
🤔